### ISSUE TITLE
Sass warning for declarations that appear after nested rules

### ISSUE THREAD

**Issue by alexgibson** [mozilla/bedrock#15117]
## Description

As of Sass 1.78.0 we're now seeing a bunch of warnings when compiling SCSS along the following lines:

```
WARNING in ./media/css/protocol/protocol-firefox.scss (./media/css/protocol/protocol-firefox.scss.webpack[javascript/auto]!=!./node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].use[1]!./node_modules/sass-loader/dist/cjs.js!./media/css/protocol/protocol-firefox.scss)
[1] Module Warning (from ./node_modules/sass-loader/dist/cjs.js):
[1] Deprecation Warning on line 96, column 4 of file:///Users/alexgibson/Documents/GitHub/bedrock/node_modules/@mozilla-protocol/core/protocol/css/components/forms/_form.scss:96:4:
[1] Sass's behavior for declarations that appear after nested
[1] rules will be changing to match the behavior specified by CSS in an upcoming
[1] version. To keep the existing behavior, move the declaration above the nested
[1] rule. To opt into the new behavior, wrap the declaration in `& {}`.
[1] 
[1] More info: https://sass-lang.com/d/mixed-decls
[1] 
[1] 96 |     font-weight: bold;
```

The docs at https://sass-lang.com/d/mixed-decls explains the issue more.

Most of the instances seem to be where we're using mixins such as `@include text-title-xs;` that include nested selectors such as media queries. We seem to do this A LOT, so we may need to change how we do this to avoid the future issue highlighted in the blog past.

**Comment by jwhitlock** [comment]
We have a similar issue on Relay after upgrading to sass 1.80. We're still on @mozilla-protocol/core 17.0.1. I came up with a solution that generates the same CSS. It may be the right solution, or you may be able to come up with something more elegant: https://github.com/mozilla/fx-private-relay/pull/5174

**Comment by alexgibson** [comment]
Thanks @jwhitlock - I filed a related Protocol issue for this https://github.com/mozilla/protocol/issues/998

**Comment by stevejalim** [comment]
@janbrasna Was this something you fixed recently?

**Comment by janbrasna** [comment]
I only silenced the thousands lines of warnings for this known issue — this ticket is to track the deprecation, which partly would get resolved in future Protocol versions, simplifying towards custom properties, but I believe we will still be left with several mixins and general utility helpers that may need rethinking in how to structure them in the source SASS order once the tooling starts to move to parity with CSS semantics of the same construct.

Especially now with the warnings silenced we need to track the issue itself, as it needs refactoring, mid-term.

**Comment by stephaniehobson** [comment]
Closing in favour of https://github.com/mozilla/protocol/issues/982

**Comment by janbrasna** [comment]
WHOA they made the change breaking like right now on minor **1.92.0**:

> "Breaking change: Emit declarations, childless at-rules, and comments in the order they appear in the source even when they're interleaved with nested rules. This obsoletes the mixed-decls deprecation."

- https://sass-lang.com/documentation/breaking-changes/mixed-decls/#the-new-way
- https://github.com/sass/dart-sass/releases/tag/1.92.0
- https://github.com/sass/sass/issues/3893#issuecomment-2248982984 _"It depends how long it takes us to get 2.0.0 ready to release. If it ends up coming together quickly, we may delay this change a bit longer to ensure users have enough time to upgrade. Note that [the Dart Sass compatibility policy](https://github.com/sass/dart-sass#compatibility-policy) means that we **may choose to make this release outside of a major version** cycle, since it's for CSS compatibility."_


That applies to all kinds of stuff like `bidi`, `details`/`summary`, `light-links`, `aspect-ratio`, `highlighted` etc. …

So just a heads up. The rules would be emitted in a different order, and the overall size rather inflated:/ **There's no way to opt-out.**

We're still at 1.91.0…
https://github.com/mozilla/bedrock/blob/7e69361dc38737ef029a82315e120da80a816bf8/package.json#L23
Consider the **next bump as breaking**.

**Comment by stephaniehobson** [comment]
Uuuuuuuuuugh.

I'm going to put a note in there to pin it to this version. I wonder if there's something I can say to dependabot to leave it alone too.

**Comment by janbrasna** [comment]
Yup: https://github.com/mozilla/bedrock/blob/7e69361dc38737ef029a82315e120da80a816bf8/.github/dependabot.yml#L11-L15

We should also constrain the version in package.json to actually be pinned; in case any random update elsewhere would still move its hash in the actual lockfile.

**Comment by janbrasna** [comment]
I also wanted to stress that the mentioned https://github.com/mozilla/protocol/issues/982 issue that all the occurrences are being linked against is just a subset of the problem — yes, moving text sizing helper mixins to more CSS-native approach is a good thing, but IIUC the issue is way more complex, e.g. as reported on (randomly selected) [picto block after atmedia with bidi](https://github.com/mozilla/protocol/blob/8c2a9cca63e2cb639bc758c60f99c925b998ec21/assets/sass/protocol/components/_picto.scss#L36-L44) and again on the following block after its mid-block `bidi()` too…:/

Basically any rule that's after a `bidi` (that inherently emits nested blocks for two selectors), no matter whether upstream in MZP, or in the local SASS here, is now split into separate CSS blocks instead of being hoisted up to the singular closure preceding the new block like in all the previous versions.

From a theoretical point of view that ensures possible cascade/inheritance interactions happen in the order it all appears marked up, but means a change from the status quo nonetheless (where we may rely on the current effect of mixins redefining the rules in the declarations that appear later — but the new order will emit them separately, below, in own closure, applying them after the mixin content block which is the breaking difference in the `mixed-decls` saga).

So this is not only the one MZP issue linked, but also all consumers like here, that will basically need either moving all block-sized mixins to be last (and/to verify that's still the intended outcome), or if kept as-is validating every single such occurrence is still rendered as expected after it's re-re-ordered back to be closer to upstream spec.
